File: templates/variants/go-connectrpc/internal/connectapi/waitlist_integration_test.go
package connectapi

import (
	"context"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"

	"connectrpc.com/connect"
	_ "github.com/jackc/pgx/v5/stdlib"

	waitlistv1 "{{MODULE_PATH}}/gen/waitlist/v1"
	waitlistv1connect "{{MODULE_PATH}}/gen/waitlist/v1/waitlistv1connect"
	"{{MODULE_PATH}}/internal/app"
)

func TestWaitlistRPCJoinIsIdempotentAndRecordsTriggers(t *testing.T) {
	databaseURL := strings.TrimSpace(os.Getenv("DATABASE_URL"))
	if databaseURL == "" {
		t.Skip("DATABASE_URL is required for integration test")
	}

	db, err := app.OpenDatabase(context.Background(), databaseURL)
	if err != nil {
		t.Fatalf("open database: %v", err)
	}
	t.Cleanup(func() { _ = db.Close() })

	if _, err := db.ExecContext(context.Background(), `
truncate table
  waitlist_triggers,
  waitlist_entries
restart identity cascade`); err != nil {
		t.Fatalf("truncate tables: %v", err)
	}

	service := app.NewWaitlistService(db)
	path, handler := NewHandler(service)
	mux := http.NewServeMux()
	mux.Handle(path, handler)
	server := httptest.NewServer(mux)
	t.Cleanup(server.Close)

	client := waitlistv1connect.NewWaitlistServiceClient(http.DefaultClient, server.URL)

	first, err := client.JoinWaitlist(context.Background(), connect.NewRequest(&waitlistv1.JoinWaitlistRequest{
		Email:   "Founder@Example.com",
		Name:    "Founder",
		Company: "Example Co",
		Source:  "homepage",
	}))
	if err != nil {
		t.Fatalf("join waitlist: %v", err)
	}
	if !first.Msg.GetCreated() {
		t.Fatal("expected first join to create entry")
	}
	if first.Msg.GetEntry().GetEmail() != "founder@example.com" {
		t.Fatalf("expected normalized email, got %s", first.Msg.GetEntry().GetEmail())
	}

	second, err := client.JoinWaitlist(context.Background(), connect.NewRequest(&waitlistv1.JoinWaitlistRequest{
		Email: "founder@example.com",
	}))
	if err != nil {
		t.Fatalf("join waitlist again: %v", err)
	}
	if second.Msg.GetCreated() {
		t.Fatal("expected second join to be idempotent")
	}
	if second.Msg.GetEntry().GetId() != first.Msg.GetEntry().GetId() {
		t.Fatalf("expected same entry id, got %s and %s", first.Msg.GetEntry().GetId(), second.Msg.GetEntry().GetId())
	}

	trigger, err := client.RecordTrigger(context.Background(), connect.NewRequest(&waitlistv1.RecordTriggerRequest{
		Type:        "cron.digest",
		EntryId:     first.Msg.GetEntry().GetId(),
		PayloadJson: "{}",
	}))
	if err != nil {
		t.Fatalf("record trigger: %v", err)
	}
	if trigger.Msg.GetTrigger().GetType() != "cron.digest" {
		t.Fatalf("expected cron.digest trigger, got %s", trigger.Msg.GetTrigger().GetType())
	}
	if trigger.Msg.GetTrigger().GetEntryId() != first.Msg.GetEntry().GetId() {
		t.Fatalf("expected entry id %s, got %s", first.Msg.GetEntry().GetId(), trigger.Msg.GetTrigger().GetEntryId())
	}

	updated, err := client.UpdateWaitlistEntry(context.Background(), connect.NewRequest(&waitlistv1.UpdateWaitlistEntryRequest{
		EntryId: first.Msg.GetEntry().GetId(),
		Status:  "invited",
	}))
	if err != nil {
		t.Fatalf("update waitlist entry: %v", err)
	}
	if updated.Msg.GetEntry().GetStatus() != "invited" {
		t.Fatalf("expected invited status, got %s", updated.Msg.GetEntry().GetStatus())
	}

	list, err := client.ListWaitlistEntries(context.Background(), connect.NewRequest(&waitlistv1.ListWaitlistEntriesRequest{
		Status: "invited",
	}))
	if err != nil {
		t.Fatalf("list waitlist entries: %v", err)
	}
	if len(list.Msg.GetEntries()) != 1 || list.Msg.GetEntries()[0].GetId() != first.Msg.GetEntry().GetId() {
		t.Fatalf("expected one invited entry %s, got %+v", first.Msg.GetEntry().GetId(), list.Msg.GetEntries())
	}

	exported, err := client.ExportWaitlistEntries(context.Background(), connect.NewRequest(&waitlistv1.ExportWaitlistEntriesRequest{
		Status: "invited",
	}))
	if err != nil {
		t.Fatalf("export waitlist entries: %v", err)
	}
	if !strings.Contains(exported.Msg.GetCsv(), "founder@example.com") {
		t.Fatalf("expected exported csv to contain email, got %s", exported.Msg.GetCsv())
	}
}

File: templates/variants/bun-connectrpc/test/waitlist.integration.test.ts
import { afterEach, beforeEach, expect, test } from "bun:test";
import { SQL } from "bun";
import { createDb } from "../src/db/client";
import { WaitlistRepository } from "../src/db/repository";
import { DefaultWaitlistService } from "../src/waitlist/service";
import { createRpcService } from "../src/index";

const databaseUrl = Bun.env.DATABASE_URL?.trim();
const integrationTest = databaseUrl ? test : test.skip;

let sql: SQL | null = null;

beforeEach(async () => {
  if (!databaseUrl) {
    return;
  }
  sql = new SQL(databaseUrl);
  await sql.unsafe(`
    truncate table
      waitlist_triggers,
      waitlist_entries
    restart identity cascade
  `);
});

afterEach(async () => {
  await sql?.end();
  sql = null;
});

integrationTest("waitlist rpc join is idempotent and records triggers", async () => {
  const rpc = createRpcService(new DefaultWaitlistService(new WaitlistRepository(createDb(databaseUrl))));

  const first = await rpc.joinWaitlist!({
    email: "Founder@Example.com",
    name: "Founder",
    company: "Example Co",
    source: "homepage",
  } as any, undefined as never);
  expect(first.created).toBe(true);
  expect(first.entry!.email).toBe("founder@example.com");

  const second = await rpc.joinWaitlist!({ email: "founder@example.com" } as any, undefined as never);
  expect(second.created).toBe(false);
  expect(second.entry!.id).toBe(first.entry!.id);

  const trigger = await rpc.recordTrigger!({
    type: "cron.digest",
    entryId: first.entry!.id,
    payloadJson: "{}",
  } as any, undefined as never);
  expect(trigger.trigger).toMatchObject({
    type: "cron.digest",
    entryId: first.entry!.id,
    status: "queued",
  });

  const updated = await rpc.updateWaitlistEntry!({
    entryId: first.entry!.id,
    status: "invited",
  } as any, undefined as never);
  expect(updated.entry).toMatchObject({ id: first.entry!.id, status: "invited" });

  const list = await rpc.listWaitlistEntries!({ status: "invited" } as any, undefined as never);
  const entries = list.entries ?? [];
  expect(entries).toHaveLength(1);
  expect(entries[0]).toMatchObject({ id: first.entry!.id, status: "invited" });

  const exported = await rpc.exportWaitlistEntries!({ status: "invited" } as any, undefined as never);
  expect(exported.csv).toContain("founder@example.com");
});
